### PR TITLE
add decode events test

### PR DESCRIPTION
**Pull request by qperrot**
Exemple of how "contract.decodeEvents" function works

### CODE REVIEW

File: test/decode-events.test.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    
+    const txHash = await contract.invoke("increase_balance", { amount: 10 });
+    const tx = await starknet.getTransaction(txHash);
+    expect(tx.status).to.be.oneOf(OK_TX_STATUSES);
+    expect(tx.transaction.calldata).to.deep.equal(["0xa"]);
+    expectAddressEquality(tx.transaction.contract_address,contract.address);
+
+    const receipt = await starknet.getTransactionReceipt(txHash);
+    const result = await contract.decodeEvents(receipt.events)
+    console.log("result: ", result)
```
Comment (FabijanC) [inline test/decode-events.test.ts:52]: Did you leave this log line on purpose? I think it can be removed.

File: test/decode-events.test.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+ * Receives a hex address, converts it to bigint, converts it back to hex.
+ * This is done to strip leading zeros.
+ * @param address a hex string representation of an address
+ * @returns an adapted hex string representation of the address
+ */
+ function adaptAddress(address: string) {
+  return "0x" + BigInt(address).toString(16);
+}
+
+const OK_TX_STATUSES = ["PENDING", "ACCEPTED_ON_L2", "ACCEPTED_ON_L1"]
```
Comment (FabijanC) [inline test/decode-events.test.ts:18]: Move this and `adaptAddress` to `util.ts`. Also modify `sample-test.ts` to import from there.

File: test/decode-events.test.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * Expects address equality after adapting them.
+ * @param actual 
+ * @param expected 
+ */
+function expectAddressEquality(actual: string, expected: string) {
```
Comment (FabijanC) [inline test/decode-events.test.ts:25]: Remove this definition and import from `util`.

File: test/decode-events.test.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    // assumes contract.cairo and events.cairo has been compiled
+    contractFactory = await starknet.getContractFactory("contract");
+    eventsContractFactory = await starknet.getContractFactory("events");
+  });
+
+  it("should decode events successfully", async function() {
+    const contract = await eventsContractFactory.deploy();
+    
+    const txHash = await contract.invoke("increase_balance", { amount: 10 });
+    const tx = await starknet.getTransaction(txHash);
```
Comment (FabijanC) [inline test/decode-events.test.ts:45]: You can remove lines 45-49, this test should only test if the decoding is done properly - assertion of getTransaction response is already done in a separate test.

File: test/decode-events.test.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+ * @param expected 
+ */
+function expectAddressEquality(actual: string, expected: string) {
+  expect(adaptAddress(actual)).to.equal(adaptAddress(expected));
+}
+
+describe("Starknet", function () {
+  this.timeout(TIMEOUT);
+
+  let contractFactory: StarknetContractFactory;
```
Comment (FabijanC) [inline test/decode-events.test.ts:32]: We don't need this `contractFactory` here, right? You can remove it.

File: test/decode-events.test.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    const contract = await eventsContractFactory.deploy();
+    
+    const txHash = await contract.invoke("increase_balance", { amount: 10 });
+    const tx = await starknet.getTransaction(txHash);
+    expect(tx.status).to.be.oneOf(OK_TX_STATUSES);
+    expect(tx.transaction.calldata).to.deep.equal(["0xa"]);
+    expectAddressEquality(tx.transaction.contract_address,contract.address);
+
+    const receipt = await starknet.getTransactionReceipt(txHash);
+    const result = await contract.decodeEvents(receipt.events)
```
Comment (FabijanC) [inline test/decode-events.test.ts:51]: This variable should be renamed from `result` to `events`.

**Review by FabijanC — CHANGES_REQUESTED** [review]

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: test/decode-events.test.ts (hunk 1 above)
Comment (qperrot) [reply test/decode-events.test.ts]: https://github.com/Shard-Labs/starknet-hardhat-example/pull/68/commits/fc8bc0d6fc5372612eac03c7054213d399366c08

File: contracts/events.cairo [hunk 7]
```diff
@@ -34,6 +39,7 @@ func increase_balance{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_ch
 
     # Emit the event.
     increase_balance_called.emit(current_balance=res, amount=amount)
+    event_test.emit(arg1=4, arg2=10, arg3=56, arg4=12)
```
[the remark this answers is not in the document]
Comment (qperrot) [reply contracts/events.cairo]: https://github.com/Shard-Labs/starknet-hardhat-example/pull/68/commits/bf08c9224dd1f823889f710b147028c2006d0012

File: contracts/events.cairo [hunk 8]
```diff
@@ -16,9 +28,13 @@ end
 func increase_balance_called(current_balance : felt, amount : felt):
 end
 
-#An event emitted to test event.
+# Events emitted to test event.
+@event
+func complex_event_test(simple : felt, struc : TestEvent, alias : AliasEvent, array_len : felt, array : felt*):
```
[the remark this answers is not in the document]
Comment (qperrot) [reply contracts/events.cairo]: https://github.com/Shard-Labs/starknet-hardhat-example/pull/68/commits/57e7f9ac6fa328cc9bc22abad97cdb01dba70c2b#diff-45358d5faa0ae92937f946f84f828609ad43f3e10e7c110576259a85bd53986b

File: test/decode-events.test.ts [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    expect(events).to.deep.equal(payload);
+  });
+
+  it("should decode events from send events successfully", async function () {
+    const txHash = await contract.invoke("send_events", {
+      array: [42, 78, 54, 8],
+    });
+    const receipt = await starknet.getTransactionReceipt(txHash);
+    const events = await contract.decodeEvents(receipt.events);
+    const test = {
```
[the remark this answers is not in the document]
Comment (qperrot) [reply test/decode-events.test.ts]: https://github.com/Shard-Labs/starknet-hardhat-example/pull/68/commits/57e7f9ac6fa328cc9bc22abad97cdb01dba70c2b#diff-9b7e3908278c95f0cdd7396c09c69a585d3a7c7051a600767c7c175867628e6b

File: test/decode-events.test.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+      a: 10n,
+      b: 45n,
+      c: 89n,
+    };
+    const payload = [
```
[the remark this answers is not in the document]
Comment (qperrot) [reply test/decode-events.test.ts]: https://github.com/Shard-Labs/starknet-hardhat-example/pull/68/commits/57e7f9ac6fa328cc9bc22abad97cdb01dba70c2b#diff-9b7e3908278c95f0cdd7396c09c69a585d3a7c7051a600767c7c175867628e6b

**Review by FabijanC — APPROVED** [review]

**Pull request merged**